File: backend/app/models/asset_hub.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class GlobalAssetFolder(Base):
    """全局资产目录。"""

    __tablename__ = "global_asset_folders"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    folder_type: Mapped[str] = mapped_column(String(50), default="generic", nullable=False)
    storage_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class GlobalCharacter(Base):
    """全局角色素材。"""

    __tablename__ = "global_characters"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    alias: Mapped[str | None] = mapped_column(String(120), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    prompt_template: Mapped[str | None] = mapped_column(Text, nullable=True)
    reference_image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    default_voice_id: Mapped[str | None] = mapped_column(
        String(36), ForeignKey("global_voices.id", ondelete="SET NULL"), nullable=True
    )
    tags_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class GlobalLocation(Base):
    """全局地点素材。"""

    __tablename__ = "global_locations"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    prompt_template: Mapped[str | None] = mapped_column(Text, nullable=True)
    reference_image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    tags_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class GlobalVoice(Base):
    """全局语音配置。"""

    __tablename__ = "global_voices"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    provider: Mapped[str] = mapped_column(String(50), default="edge-tts", nullable=False)
    voice_code: Mapped[str] = mapped_column(String(200), nullable=False)
    language: Mapped[str | None] = mapped_column(String(30), nullable=True)
    gender: Mapped[str | None] = mapped_column(String(30), nullable=True)
    sample_audio_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    style_prompt: Mapped[str | None] = mapped_column(Text, nullable=True)
    meta_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

    panels: Mapped[list["Panel"]] = relationship("Panel", back_populates="voice")

File: frontend/src/types/assetHub.ts
export interface AssetFolder {
  id: string
  name: string
  folder_type: string
  storage_path: string | null
  description: string | null
  sort_order: number
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface GlobalVoice {
  id: string
  name: string
  project_id: string | null
  provider: string
  voice_code: string
  folder_id: string | null
  language: string | null
  gender: string | null
  sample_audio_url: string | null
  style_prompt: string | null
  meta: Record<string, unknown>
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface GlobalCharacterAsset {
  id: string
  name: string
  project_id: string | null
  folder_id: string | null
  alias: string | null
  description: string | null
  prompt_template: string | null
  reference_image_url: string | null
  default_voice_id: string | null
  tags: string[]
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface GlobalLocationAsset {
  id: string
  name: string
  project_id: string | null
  folder_id: string | null
  description: string | null
  prompt_template: string | null
  reference_image_url: string | null
  tags: string[]
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface AssetHubOverview {
  folders: AssetFolder[]
  characters: GlobalCharacterAsset[]
  locations: GlobalLocationAsset[]
  voices: GlobalVoice[]
}
